File: backend/consequence_engine/propagation.py
"""
THE NARRATIVE — CONSEQUENCE PROPAGATION ENGINE (CPE) v2 — server-side core.

Authoritative, server-side port of web/src/lib/propagation.js. Keeping the engine
(and its tuned parameters) on the server is an IP requirement: the constants below
are the trade-secret surface and must never ship to the browser. The client should
consume only computed scores + drivers via /api/v1/exposure.

Deterministic. Explainable. Runs WITHOUT the LLM. See docs/IP-AND-ALGORITHMS.md.

base(e)       = importance(e) · confidence(e) · recency(e)
effective(e)  = base(e) + LAMBDA · Σ_{c→e} w · effective(c)         (Katz diffusion)
shock(entity) = Σ_e Σ_impacts sign · severity · effective(e)
ExposureIndex = 100 · (1 − e^(−max(net_shock, 0) / K))              (bounded 0–100)
"""

from __future__ import annotations

import math
import re
import time
from datetime import datetime

from backend.consequence_engine.corroboration import CORROB_W

# Tuned "secret sauce" — versioned with the model. SERVER-SIDE ONLY.
ENGINE_VERSION = "2.0"
LAMBDA = 0.5            # per-hop causal decay (LAMBDA·ρ(W) < 1 ⇒ Katz series converges)
K = 0.8                # exposure saturation constant
INDIRECT_FACTOR = 0.6  # indirect impacts count less than direct
SEVERITY = {"critical": 1.0, "high": 0.72, "medium": 0.45, "low": 0.22}
EVIDENCE = {"VERIFIED": 1.0, "INFERRED": 0.6, "SPECULATIVE": 0.3}
KAPPA = 2.5            # corroboration saturation (independent sources)
DELTA = 0.5           # max confidence penalty from fully-disputed claims
TAU_HOURS = 168.0     # event-freshness e-folding horizon (≈7 days)
MAX_ITERS = 64        # Katz diffusion iteration cap
EPSILON = 1e-4        # Katz convergence tolerance
K_EVENT = 1.2         # saturation for per-event exposure heat
DISRUPTION_K = 0.8    # max shock from traffic disruption near an event
TAU_TRAFFIC = 30.0    # traffic-count saturation for disruption emissions
MARKET_W = 0.5        # how much per-sector market stress amplifies its exposure

ENTITY_ALIASES = {
    "shipping and logistics": "shipping",
    "logistics": "shipping",
    "maritime shipping": "shipping",
    "oil and gas": "energy",
    "fuel": "energy",
    "consumer prices": "consumer goods",
    "semiconductor supply": "semiconductors",
    "tech industry": "technology",
    "ai": "technology",
}

_PARAMS_PUBLIC = {
    "version": ENGINE_VERSION,
    # NOTE: only the version is public. The constants are intentionally omitted.
}


def _clamp01(x: float) -> float:
    return max(0.0, min(1.0, x))


def _as_list(v):
    if isinstance(v, list):
        return v
    return [v] if v else []


def _sev_weight(s) -> float:
    return SEVERITY.get(str(s or "medium").lower(), SEVERITY["medium"])


def _importance_of(e: dict) -> float:
    raw = e.get("importance_score")
    if raw is None:
        raw = e.get("global_importance_score")
    if raw is None:
        raw = e.get("importance", 0)
    return _clamp01((raw or 0) / 100.0)


def canonicalize(name) -> str:
    s = re.sub(r"\s+", " ", re.sub(r"[^a-z0-9 ]+", " ", str(name or "").lower().replace("&", " and "))).strip()
    return ENTITY_ALIASES.get(s, s)


def _to_ms(raw) -> float | None:
    if raw is None:
        return None
    if isinstance(raw, datetime):
        return raw.timestamp() * 1000.0
    try:
        s = str(raw).replace("Z", "+00:00")
        return datetime.fromisoformat(s).timestamp() * 1000.0
    except ValueError:
        return None


def _recency_of(e: dict, now_ms: float) -> float:
    ts = _to_ms(e.get("last_updated_at") or e.get("first_detected_at") or e.get("created_at"))
    if ts is None:
        return 1.0
    age_hours = max(0.0, (now_ms - ts) / 3_600_000.0)
    return math.exp(-age_hours / TAU_HOURS)


def _confidence_of(e: dict) -> float:
    m = e.get("consequence_map") or {}
    chain = m.get("consequence_chain") or []
    chain = chain if isinstance(chain, list) else []

    if chain:
        s = 0.0
        for n in chain:
            t = str((n or {}).get("type", "")).upper()
            s += EVIDENCE["VERIFIED"] if "VERIFIED" in t else EVIDENCE["INFERRED"] if "INFERRED" in t else EVIDENCE["SPECULATIVE"]
        grade = s / len(chain)
    else:
        c = m.get("confidence")
        grade = 0.85 if c == "high" else 0.4 if c == "low" else 0.6

    n_sources = len(m.get("sources_analyzed") or []) or e.get("sources_count") or len(e.get("articles") or []) or (len(chain) or 1)
    corroboration = 1 - math.exp(-n_sources / KAPPA)

    disputed = len(m.get("disputed_points") or [])
    dispute_penalty = DELTA * (disputed / (len(chain) + disputed)) if disputed else 0.0

    return _clamp01(grade * corroboration * (1 - dispute_penalty))


def _base_of(e: dict, now_ms: float) -> float:
    base = _importance_of(e) * _confidence_of(e) * _recency_of(e, now_ms)
    # Cross-feed corroboration uplift (default 0 ⇒ no-op): independent feeds
    # converging on the same place+time make an event's shock more credible.
    idx = e.get("corroboration_index")
    if idx:
        base *= 1 + CORROB_W * _clamp01(idx)
    return base


_MITIGATE_RE = re.compile(r"mitigat|reduc|de-?escalat|resolv")


def _is_mitigating(i: dict) -> bool:
    return bool(_MITIGATE_RE.search(str((i or {}).get("direction") or (i or {}).get("effect") or "").lower()))


def _sector_emissions(e: dict, traffic_by_event: dict | None = None):
    m = e.get("consequence_map") or {}
    out = []
    for i in _as_list(m.get("direct_impact")):
        if i and i.get("sector"):
            out.append((i["sector"].strip(), _sev_weight(i.get("severity")), -1 if _is_mitigating(i) else 1))
    for i in _as_list(m.get("indirect_impact")):
        if i and i.get("sector"):
            out.append((i["sector"].strip(), _sev_weight(i.get("severity")) * INDIRECT_FACTOR, -1 if _is_mitigating(i) else 1))
    tr = (traffic_by_event or {}).get(str(e["id"]))
    if tr:
        esc = 1.0 if e.get("current_status") == "escalating" else 0.5
        disrupt = lambda n: DISRUPTION_K * (1 - math.exp(-(n or 0) / TAU_TRAFFIC)) * esc  # noqa: E731
        if tr.get("vessels"):
            out.append(("shipping", disrupt(tr["vessels"]), 1))
        if tr.get("aircraft"):
            out.append(("aviation", disrupt(tr["aircraft"]), 1))
    return out


def _region_emissions(e: dict):
    geos = e.get("geography") or e.get("geographic_relevance") or []
    m = e.get("consequence_map") or {}
    impacts = _as_list(m.get("direct_impact")) + _as_list(m.get("indirect_impact"))
    max_sev = max((_sev_weight(i.get("severity")) for i in impacts), default=0.5)
    all_mit = len(impacts) > 0 and all(_is_mitigating(i) for i in impacts)
    return [(str(g).strip(), max_sev, -1 if all_mit else 1) for g in geos]


def _diffuse(events: list[dict], edges: list[dict], now_ms: float):
    by_id = {str(e["id"]): e for e in events}
    ids = [str(e["id"]) for e in events]

    in_adj: dict[str, list[tuple[str, float]]] = {i: [] for i in ids}
    for ed in edges or []:
        a = str(ed.get("source_event_id", ed.get("source")))
        b = str(ed.get("target_event_id", ed.get("target")))
        w = ed.get("weight", ed.get("connection_weight", 0.5)) or 0.5
        if a not in by_id or b not in by_id:
            continue
        in_adj[b].append((a, w))
        if ed.get("directed") is not True:
            in_adj[a].append((b, w))

    base = {i: _base_of(by_id[i], now_ms) for i in ids}
    eff = dict(base)
    attr: dict[str, dict[str, float]] = {i: {i: base[i]} for i in ids}

    for _ in range(MAX_ITERS):
        next_eff: dict[str, float] = {}
        next_attr: dict[str, dict[str, float]] = {}
        max_delta = 0.0
        for i in ids:
            v = base[i]
            m: dict[str, float] = {i: base[i]}
            for frm, w in in_adj[i]:
                v += LAMBDA * w * eff[frm]
                for src, amt in attr[frm].items():
                    m[src] = m.get(src, 0.0) + LAMBDA * w * amt
            next_eff[i] = v
            next_attr[i] = m
            max_delta = max(max_delta, abs(v - eff[i]))
        eff, attr = next_eff, next_attr
        if max_delta < EPSILON:
            break

    return by_id, attr


def _propagate(events, edges, emit, now_ms):
    by_id, attr = _diffuse(events, edges, now_ms)

    contrib: dict[str, dict] = {}
    for e in events:
        eid = str(e["id"])
        eff_attr = attr[eid]
        for target, w, sign in emit(e):
            key = canonicalize(target)
            if not key:
                continue
            slot = contrib.setdefault(key, {"display": target, "parts": {}})
            for src, amt in eff_attr.items():
                slot["parts"][src] = slot["parts"].get(src, 0.0) + sign * w * amt

    def title_of(i):
        ev = by_id.get(i)
        return (ev or {}).get("canonical_title") or (ev or {}).get("title") or i

    out = []
    for key, slot in contrib.items():
        pos = sum(a for a in slot["parts"].values() if a > 0)
        neg = sum(-a for a in slot["parts"].values() if a < 0)
        net = pos - neg
        score = round(100 * (1 - math.exp(-max(net, 0) / K)))

        ordered = sorted(slot["parts"].items(), key=lambda kv: (-kv[1], str(kv[0])))
        drivers = [
            {"id": i, "title": title_of(i), "category": (by_id.get(i) or {}).get("category"), "pct": round(100 * amt / pos) if pos else 0}
            for i, amt in ordered if amt > 0
        ][:3]
        mitigators = [
            {"id": i, "title": title_of(i), "category": (by_id.get(i) or {}).get("category"), "pct": round(100 * -amt / neg) if neg else 0}
            for i, amt in sorted(slot["parts"].items(), key=lambda kv: kv[1]) if amt < 0
        ][:2]

        out.append({"name": slot["display"], "key": key, "score": score, "net": round(net, 4), "raw": round(pos, 4), "drivers": drivers, "mitigators": mitigators})

    # Per-event drive: total exposure each event projects across all entities.
    event_drive: dict[str, float] = {}
    for slot in contrib.values():
        for src, amt in slot["parts"].items():
            event_drive[src] = event_drive.get(src, 0.0) + abs(amt)

    out.sort(key=lambda x: (-x["score"], x["key"]))
    return out, event_drive


def combine_stress(*stresses: dict | None) -> dict:
    """Merge several {sector: 0-1} stress dicts into one (max per sector, clamped).

    Lets independent stress sources — market moves, chokepoint congestion, space
    weather — share the single market_stress channel without double-counting.
    """
    out: dict[str, float] = {}
    for s in stresses:
        for sec, raw in (s or {}).items():
            try:
                v = max(0.0, min(1.0, float(raw)))
            except (TypeError, ValueError):
                continue
            if v > out.get(sec, 0.0):
                out[sec] = v
    return out


def compute_exposure_model(events: list[dict], edges: list[dict] | None = None, now_ms: float | None = None,
                           traffic_by_event: dict | None = None, market_stress: dict | None = None,
                           corroboration: dict | None = None) -> dict:
    """Full exposure model from the live event graph. Public-safe (no params leaked).

    `corroboration` maps event id → {"index": 0-1} (from corroboration.corroborate);
    a corroborated event's base shock is amplified before diffusion.
    """
    if now_ms is None:
        now_ms = time.time() * 1000.0
    edges = edges or []

    if corroboration:
        def _corrob_idx(e):
            rec = corroboration.get(str(e.get("id")))
            return rec.get("index", 0.0) if isinstance(rec, dict) else (rec or 0.0)
        # copy (never mutate the caller's events) so no index leaks across calls
        events = [{**e, "corroboration_index": _corrob_idx(e)} for e in events]
    sectors, s_drive = _propagate(events, edges, lambda e: _sector_emissions(e, traffic_by_event), now_ms)
    regions, r_drive = _propagate(events, edges, _region_emissions, now_ms)

    # Market stress amplifies exposure on stressed sectors (bounded, transparent).
    if market_stress:
        idx = {x["key"]: x for x in sectors}
        for sec, raw in market_stress.items():
            key = canonicalize(sec)
            st = max(0.0, min(1.0, raw))
            if key in idx:
                x = idx[key]
                x["score"] = round(x["score"] + (100 - x["score"]) * st * MARKET_W)
            elif st > 0:
                sectors.append({"name": sec, "key": key, "score": round(100 * st * MARKET_W),
                                "net": 0.0, "raw": 0.0, "drivers": [], "mitigators": [], "market": True})
        sectors.sort(key=lambda x: (-x["score"], x["key"]))

    drive: dict[str, float] = {}
    for d in (s_drive, r_drive):
        for k, v in d.items():
            drive[k] = drive.get(k, 0.0) + v
    event_scores = {k: round(100 * (1 - math.exp(-v / K_EVENT))) for k, v in drive.items()}

    top = sectors[:5]
    pressure = round(sum(x["score"] for x in top) / len(top)) if top else 0
    return {
        "sectors": sectors,
        "regions": regions,
        "event_scores": event_scores,
        "pressure": pressure,
        "meta": {"events": len(events), "links": len(edges), **_PARAMS_PUBLIC},
    }


def profile_exposure(profile: dict, model: dict) -> dict:
    """Personalised exposure for a user profile of sectors/regions, with attribution."""
    wanted = {canonicalize(s) for s in (profile.get("sectors") or []) + (profile.get("regions") or [])}
    matches = [x for x in (model["sectors"] + model["regions"]) if x["key"] in wanted]
    if not matches:
        return {"score": 0, "drivers": []}
    mx = max(m["score"] for m in matches)
    mean = sum(m["score"] for m in matches) / len(matches)
    score = round(0.65 * mx + 0.35 * mean)
    agg: dict[str, dict] = {}
    for m in matches:
        for d in m["drivers"]:
            slot = agg.setdefault(d["id"], {**d, "pct": 0})
            slot["pct"] += d["pct"]
    drivers = sorted(agg.values(), key=lambda d: -d["pct"])[:3]
    return {"score": score, "drivers": drivers}

File: web/src/lib/propagation.js
// ─────────────────────────────────────────────────────────────────────────────
//  THE NARRATIVE — CONSEQUENCE PROPAGATION ENGINE (CPE)  v2
//  Proprietary core. Deterministic. Explainable. Runs WITHOUT the LLM.
//
//  The LLM (consensus_mapper) is a *feeder*: it extracts evidence-graded impacts
//  per event. This engine is the product — it assembles a typed causal graph and
//  propagates each event's shock outward to compute a bounded, attributable
//  EXPOSURE INDEX for any entity (sector, region, commodity, or user profile).
//
//  v2 upgrades over v1 (all still deterministic + explainable + attributable):
//   • n-hop influence via a converging Katz/personalized-PageRank diffusion
//     (replaces the single fixed 2-hop term) — true multi-step causal reach.
//   • Confidence now rewards corroboration (# independent sources, saturating)
//     and penalizes the LLM's own `disputed_points` (previously ignored).
//   • Temporal decay — fresh shocks dominate; stale events fade (e-folding TAU).
//   • Signed / mitigating impacts — a ceasefire or resolution REDUCES exposure,
//     so the net index is genuinely meaningful (not monotonic-up only).
//   • Entity resolution — canonical alias map collapses "Shipping & Logistics"
//     and "Shipping" so exposure stops silently fragmenting across name variants.
//   • Directed edges — when the graph carries cause→effect direction, influence
//     flows downstream only; otherwise it degrades gracefully to undirected.
//   • Full attribution survives the diffusion: every entity's score decomposes
//     into ranked SOURCE events (drivers) + mitigating factors, with % share.
//
//  Math
//  ────
//  base(e)        = importance(e) · confidence(e) · recency(e)
//  effective(e)   = base(e) + LAMBDA · Σ_{c→e}  w(c→e) · effective(c)   (Katz)
//  shock(entity)  = Σ_e  Σ_impacts  sign · severity · effective(e)
//  ExposureIndex  = 100 · (1 − e^(−max(net_shock,0) / K))               (0–100)
//
//  Why this is defensible IP (not prompt engineering):
//   • a named, deterministic algorithm with explicit, tunable, versioned params
//   • explainable end-to-end (every score decomposes to source events)
//   • confidence-weighted by evidence grade (VERIFIED > INFERRED > SPECULATIVE)
//   • compounds with proprietary data (more events/edges/outcomes → sharper graph)
// ─────────────────────────────────────────────────────────────────────────────

// NOTE (IP): this module is the trade-secret surface (the engine + its tuned
// constants). It is the AUTHORITATIVE engine ONLY server-side
// (backend/consequence_engine/propagation.py). This client copy exists solely to
// render the OFFLINE DEMO and must never be statically imported — load it lazily
// behind a DEMO_MODE guard so it is dead-code-eliminated from production builds.
// See docs/IP-AND-ALGORITHMS.md.
import { canonicalize } from "./exposureProfile.js";

// Engine parameters (the tuned "secret sauce" — versioned with the model).
export const CPE = {
  VERSION: "2.0",
  ENGINE_VERSION: "2.0",
  LAMBDA: 0.5,            // per-hop decay for causal amplification (LAMBDA·ρ(W) < 1 ⇒ converges)
  K: 0.8,                // saturation constant for the exposure curve
  INDIRECT_FACTOR: 0.6,  // downstream (indirect) impacts count less than direct
  SEVERITY: { critical: 1.0, high: 0.72, medium: 0.45, low: 0.22 },
  EVIDENCE: { VERIFIED: 1.0, INFERRED: 0.6, SPECULATIVE: 0.3 },
  KAPPA: 2.5,            // corroboration saturation — sources needed for high confidence
  DELTA: 0.5,            // max confidence penalty from fully-disputed claims
  TAU_HOURS: 168,        // event-freshness e-folding horizon (≈7 days)
  MAX_ITERS: 64,         // Katz diffusion iteration cap
  EPSILON: 1e-4,         // Katz convergence tolerance (max per-node delta)
  K_EVENT: 1.2,          // saturation for per-event exposure heat
  DISRUPTION_K: 0.8,     // max shock from traffic disruption near an event
  TAU_TRAFFIC: 30,       // traffic-count saturation for disruption emissions
};

const clamp01 = (x) => Math.max(0, Math.min(1, x));
const asArray = (v) => (Array.isArray(v) ? v : v ? [v] : []);
const sevWeight = (s) => CPE.SEVERITY[String(s || "medium").toLowerCase()] ?? CPE.SEVERITY.medium;
const importanceOf = (e) => clamp01((e.importance_score ?? e.global_importance_score ?? e.importance ?? 0) / 100);

// Freshness: events decay with age so a 6-month-old shock no longer dominates.
function recencyOf(e, now) {
  const raw = e.last_updated_at || e.first_detected_at || e.created_at || e.detected_at;
  if (!raw) return 1; // no timestamp → treat as current (don't punish mock/test data)
  const t = new Date(raw).getTime();
  if (Number.isNaN(t)) return 1;
  const ageHours = Math.max(0, (now - t) / 3_600_000);
  return Math.exp(-ageHours / CPE.TAU_HOURS);
}

// Confidence from evidence grade × corroboration × (1 − dispute). The LLM feed's
// evidence grades, source count, and disputed points all flow in here.
function confidenceOf(e) {
  const m = e.consequence_map || {};
  const chain = m.consequence_chain || [];

  let grade;
  if (chain.length) {
    let s = 0;
    for (const n of chain) {
      const t = (n.type || "").toUpperCase();
      s += t.includes("VERIFIED") ? CPE.EVIDENCE.VERIFIED
        : t.includes("INFERRED") ? CPE.EVIDENCE.INFERRED
        : CPE.EVIDENCE.SPECULATIVE;
    }
    grade = s / chain.length;
  } else {
    const c = m.confidence;
    grade = c === "high" ? 0.85 : c === "low" ? 0.4 : 0.6;
  }

  const nSources = (m.sources_analyzed?.length)
    ?? e.sources_count
    ?? (e.articles?.length)
    ?? (chain.length || 1);
  const corroboration = 1 - Math.exp(-nSources / CPE.KAPPA);

  const disputed = m.disputed_points?.length || 0;
  const disputePenalty = disputed ? CPE.DELTA * (disputed / (chain.length + disputed)) : 0;

  return clamp01(grade * corroboration * (1 - disputePenalty));
}

// The seed shock each event injects before any propagation.
const baseOf = (e, now) => importanceOf(e) * confidenceOf(e) * recencyOf(e, now);

// Whether an impact mitigates (reduces) rather than aggravates exposure.
const isMitigating = (i) =>
  String(i?.direction || i?.effect || "").toLowerCase().match(/mitigat|reduc|de-?escalat|resolv/) != null;

// Entity emitters — what each event projects shock onto, how hard, and which sign.
// `trafficByEvent` (optional) injects live disruption: heavy ship/plane traffic near
// an event raises Shipping / Aviation exposure — traffic disruption as a consequence.
function sectorEmissions(e, trafficByEvent) {
  const m = e.consequence_map || {};
  const out = [];
  asArray(m.direct_impact).forEach((i) => {
    if (i?.sector) out.push({ target: i.sector.trim(), w: sevWeight(i.severity), sign: isMitigating(i) ? -1 : 1 });
  });
  asArray(m.indirect_impact).forEach((i) => {
    if (i?.sector) out.push({ target: i.sector.trim(), w: sevWeight(i.severity) * CPE.INDIRECT_FACTOR, sign: isMitigating(i) ? -1 : 1 });
  });
  const tr = trafficByEvent && trafficByEvent[String(e.id)];
  if (tr) {
    const esc = e.current_status === "escalating" ? 1 : 0.5;
    const disrupt = (n) => CPE.DISRUPTION_K * (1 - Math.exp(-(n || 0) / CPE.TAU_TRAFFIC)) * esc;
    if (tr.vessels) out.push({ target: "shipping", w: disrupt(tr.vessels), sign: 1 });
    if (tr.aircraft) out.push({ target: "aviation", w: disrupt(tr.aircraft), sign: 1 });
  }
  return out;
}

function regionEmissions(e) {
  const geos = e.geography || e.geographic_relevance || [];
  const impacts = [...asArray(e.consequence_map?.direct_impact), ...asArray(e.consequence_map?.indirect_impact)];
  const maxSev = impacts.length ? Math.max(...impacts.map((i) => sevWeight(i.severity))) : 0.5;
  // Region sign follows the dominant impact: an all-mitigating event eases regional exposure.
  const allMitigating = impacts.length > 0 && impacts.every(isMitigating);
  return geos.map((g) => ({ target: String(g).trim(), w: maxSev, sign: allMitigating ? -1 : 1 }));
}

// ── Katz/PPR diffusion over the event→event graph, with per-source attribution ─
// effective(e) = base(e) + LAMBDA·Σ_{c→e} w·effective(c), iterated to convergence.
// attr.get(e) decomposes effective(e) into the SOURCE events it traces back to.
function diffuse(events, edges, now) {
  const byId = new Map(events.map((e) => [String(e.id), e]));
  const ids = events.map((e) => String(e.id));

  // Incoming adjacency: in.get(effect) = [{ from: cause, w }]
  const inAdj = new Map(ids.map((id) => [id, []]));
  (edges || []).forEach((ed) => {
    const a = String(ed.source_event_id ?? ed.source);
    const b = String(ed.target_event_id ?? ed.target);
    const w = ed.weight ?? ed.connection_weight ?? 0.5;
    if (!byId.has(a) || !byId.has(b)) return;
    inAdj.get(b).push({ from: a, w });                 // cause a → effect b
    if (ed.directed !== true) inAdj.get(a).push({ from: b, w }); // undirected ⇒ both ways
  });

  const base = new Map(ids.map((id) => [id, baseOf(byId.get(id), now)]));
  let eff = new Map(base);
  let attr = new Map(ids.map((id) => [id, new Map([[id, base.get(id)]])]));

  for (let iter = 0; iter < CPE.MAX_ITERS; iter++) {
    const nextEff = new Map();
    const nextAttr = new Map();
    let maxDelta = 0;

    for (const id of ids) {
      let v = base.get(id);
      const m = new Map([[id, base.get(id)]]);
      for (const { from, w } of inAdj.get(id)) {
        v += CPE.LAMBDA * w * eff.get(from);
        for (const [src, amt] of attr.get(from)) {
          m.set(src, (m.get(src) || 0) + CPE.LAMBDA * w * amt);
        }
      }
      nextEff.set(id, v);
      nextAttr.set(id, m);
      maxDelta = Math.max(maxDelta, Math.abs(v - eff.get(id)));
    }

    eff = nextEff;
    attr = nextAttr;
    if (maxDelta < CPE.EPSILON) break;
  }

  return { byId, attr };
}

// Core propagation over a given emitter. Returns ranked entities with drivers.
function propagate(events, edges, emit, now) {
  const { byId, attr } = diffuse(events, edges, now);

  // entity(canonical) -> { display, parts: Map<sourceEventId, signedAmount> }
  const contrib = new Map();
  for (const e of events) {
    const eid = String(e.id);
    const effAttr = attr.get(eid);
    for (const { target, w, sign } of emit(e)) {
      const key = canonicalize(target);
      if (!key) continue;
      const slot = contrib.get(key) || { display: target, parts: new Map() };
      for (const [src, amt] of effAttr) {
        slot.parts.set(src, (slot.parts.get(src) || 0) + sign * w * amt);
      }
      contrib.set(key, slot);
    }
  }

  const titleOf = (id) => byId.get(id)?.canonical_title || byId.get(id)?.title || id;
  const out = [];
  for (const [key, slot] of contrib) {
    let pos = 0, neg = 0;
    for (const amt of slot.parts.values()) { if (amt > 0) pos += amt; else neg += -amt; }
    const net = pos - neg;
    const score = Math.round(100 * (1 - Math.exp(-Math.max(net, 0) / CPE.K)));

    const sorted = [...slot.parts.entries()].sort((a, b) => (b[1] - a[1]) || String(a[0]).localeCompare(String(b[0])));
    const drivers = sorted.filter(([, amt]) => amt > 0).slice(0, 3).map(([id, amt]) => ({
      id, title: titleOf(id), category: byId.get(id)?.category, pct: pos ? Math.round((100 * amt) / pos) : 0,
    }));
    const mitigators = sorted.filter(([, amt]) => amt < 0).sort((a, b) => a[1] - b[1]).slice(0, 2).map(([id, amt]) => ({
      id, title: titleOf(id), category: byId.get(id)?.category, pct: neg ? Math.round((100 * -amt) / neg) : 0,
    }));

    out.push({ name: slot.display, key, score, net, raw: pos, drivers, mitigators });
  }

  // Per-event drive: total exposure this event projects across all entities.
  const eventDrive = new Map();
  for (const slot of contrib.values()) {
    for (const [src, amt] of slot.parts) eventDrive.set(src, (eventDrive.get(src) || 0) + Math.abs(amt));
  }

  return { ranked: out.sort((a, b) => (b.score - a.score) || a.key.localeCompare(b.key)), eventDrive };
}

/**
 * Compute the full exposure model from the live event graph.
 * @param {Array} events  events with importance + consequence_map + geography
 * @param {Array} edges   event→event causal connections (optionally `directed`)
 * @param {{now?: number, trafficByEvent?: Object}} [opts]  fixed clock + live traffic disruption
 * @returns {{sectors, regions, pressure, eventScores, meta}}
 */
export function computeExposureModel(events = [], edges = [], opts = {}) {
  const now = opts.now ?? Date.now();
  const tbe = opts.trafficByEvent;
  const s = propagate(events, edges, (e) => sectorEmissions(e, tbe), now);
  const r = propagate(events, edges, regionEmissions, now);
  const sectors = s.ranked, regions = r.ranked;

  // Per-event exposure heat (0–100): how much consequence each event drives.
  const drive = new Map();
  for (const m of [s.eventDrive, r.eventDrive]) for (const [id, v] of m) drive.set(id, (drive.get(id) || 0) + v);
  const eventScores = {};
  for (const [id, v] of drive) eventScores[id] = Math.round(100 * (1 - Math.exp(-v / CPE.K_EVENT)));

  const top = sectors.slice(0, 5);
  const pressure = top.length ? Math.round(top.reduce((acc, x) => acc + x.score, 0) / top.length) : 0;
  return {
    sectors,
    regions,
    eventScores,
    pressure,
    meta: {
      events: events.length,
      links: (edges || []).length,
      version: CPE.VERSION,
      params: { LAMBDA: CPE.LAMBDA, K: CPE.K, INDIRECT_FACTOR: CPE.INDIRECT_FACTOR, TAU_HOURS: CPE.TAU_HOURS, KAPPA: CPE.KAPPA, DELTA: CPE.DELTA },
    },
  };
}
